webrtc/webrtc.md

```diff
@@ -0,0 +1,121 @@
+|-----------------|---------------|--------|-----------------|
+| 1A              | Working Draft | Active | r0, 2023-03-03  |
+
+Authors: [@mxinden]
+
+## Motivation
+
+libp2p transport protocol enabling two private nodes (e.g. two browsers) to establish a direct connection.
+
+Browser _A_ wants to connect to Browser node _B_ with the help of server node _R_.
+Both _A_ and _B_ can not listen for incoming connections due to running in a constrained environment (i.e. a browser) with its only transport capability being the W3C WebRTC `RTCPeerConnection` API and being behind a NAT and/or firewall.
+Note that _A_ and/or _B_ may as well be non-browser nodes behind NATs and/or firewalls.
+However, for two non-browser nodes using TCP or QUIC hole punching with [DCUtR] will be the more efficient way to establish a direct connection.
+
+On a historical note, this specification replaces the existing [libp2p WebRTC star](https://github.com/libp2p/js-libp2p-webrtc-star) and [libp2p WebRTC direct](https://github.com/libp2p/js-libp2p-webrtc-direct) protocols.
+
+## Connection Establishment
+
+1. _B_ advertises support for the WebRTC browser-to-browser protocol by appending `/webrtc` to its relayed multiaddr e.g. `/ip6/fe80::883:a581:fff1:833/udp/4001/quic/webtransport/certhash/<hash1>/p2p/<relay-peer-id>/p2p-circuit/webrtc/p2p/<b-peer-id>`.
+
+2. Upon discovery of _B_'s multiaddress, _A_ learns that _B_ supports the WebRTC transport through the (constrained) W3C API and knows how to establish a relayed connection to _B_ to run the `/webrtc-signaling` protocol on top.
+
+3. _A_ establish a relayed connection to _B_.
+   Note that further steps depend on the relayed connection to be authenticated, i.e. that data sent on the relayed connection can be trusted.
+
+4. _A_ (outbound side of relayed connection) creates an `RTCPeerConnection` provided by a W3C compliant WebRTC implementation (e.g. a browser).
+   See [STUN](#stun) section on what STUN servers to configure at creation time.
+   _A_ creates an SDP offer via `RTCPeerConnection.createOffer()`.
+   _A_ initiates the signaling protocol to _B_ via the relayed connection from (1), see [Signaling Protocol](#signaling-protocol) and sends the offer to _B_.
+
+5. _B_ (inbound side of relayed connection) creates an `RTCPeerConnection`.
+   Again see [STUN](#stun) section on what STUN servers to configure at creation time.
+   _B_ receives _A_'s offer sent in (2) via the signaling protocol stream and provides the offer to its `RTCPeerConnection` via `RTCPeerConnection.setRemoteDescription`.
+   _B_ then creates an answer via `RTCPeerConnection.createAnswer` and sends it to _A_ via the existing signaling protocol stream (see [Signaling Protocol](#signaling-protocol)).
+
+6. _A_ receives _B_'s answer via the signaling protocol stream and sets it locally via `RTCPeerConnection.setRemoteDescription`.
+
+7. _A_ and _B_ send their local ICE candidates via the existing signaling protocol stream.
```

Comment:
Should we mention that this is using trickling? Is there a browser option that needs to be set to enable trickling or do browsers just do it automatically?

Reply:
Line 42 says that we add them as they come in. That is the process of trickling right? Not sure it needs an extra mention.

Maybe we can say:

```suggestion
7. _A_ and _B_ send their local ICE candidates via the existing signaling protocol stream to enable trickle ICE.
```

That would make clear that we are doing this for the purpose of trickling.

Reply:
Sounds good. Added via fca22b3.